### DISCUSSION TITLE
Importing maps of stochastic scenarios and applying stochastic scenarios to units and connections.

### DISCUSSION THREAD

**Discussion by SeanByrne08** [spine-tools/SpineOpt.jl#767] [Support]
Hi, 

I am experiencing two problems, the first is regarding importing a map which contains different time series. The second is regarding how I connect my stochastic structure to units and connections. 

Question 1: 
I am trying to import a map for the demand at a node. I have attached an image of my attempt at mapping it through the importer. The map I want to import contains 4 time series. In the mapping specification I can input in 4 different parameter indexes but only one parameter value. My time series share a parameter index and all have different parameter values. I tried to input in the values as indexes but this did not work.
![Importer mapping image](https://github.com/spine-tools/SpineOpt.jl/assets/141043201/0ad74c26-6adb-41ec-9fe1-dab98138f7b0)

Question 2: 
I have applied this stochastic scenarios to a node by inputting the data directly into the DB editor. I would also like to connect the same stochastic structure to different units and connections. I have tried using  the available unit/connection to stochastic structure relationship classes the code runs into errors when I do.

**Comment by tarskul** [comment]
I do not fully understand what you want to do but maybe this can help in the meanwhile:

1. start from the simple system tutorial
2. add a scenario object
3. add a connection between the stochastic structure and the scenario object

![image](https://github.com/spine-tools/SpineOpt.jl/assets/120471470/81f9e922-7a6a-4b7b-85b2-7b25f7103bdf)

4. change the demand value to a map with the keys corresponding to the scenarios and the values 150 and 100

![image](https://github.com/spine-tools/SpineOpt.jl/assets/120471470/296a9da4-e7bc-4db2-80e3-35164865e451)

5. in the output you will see a value for the unit flow from each plant in each scenario, e.g. 50 and 0 respectively for plant b

That should result in a working stochastic structure and perhaps you can figure out what you want to do from there.

**Reply by SeanByrne08** [reply]
Hi, 

Thank you for your response! I have managed to add stochastic scenarios, like in your example, to nodes. However, i have not been able to apply then to units or connections.

The options to connect a stochastic structure to a connection is just connection_investment_stochatsic_structure.
The options to connect a stochastic structure to a unit are unit_investment_stochatsic_structure and unit_on_stochastic_structure.

None of these option have worked for me when applying the same steps as you laid out above. 

The other part of my problem was how to use the importer to import in a map.

**Comment by tarskul** [comment]
For a simple stochastic structure it is sufficient to define your scenarios globally as before and use the maps similar to the node, i.e. you do not need to connect them with connection_investment_stochastic_structure, e.g. in the example from before you could also just make a map for the vom_cost.

![image](https://github.com/spine-tools/SpineOpt.jl/assets/120471470/9dabbbe7-9ca5-4a90-80cd-3ed25258f234)

For more complex structures with different structures in different parts of the system you need to know that everything needs to be part of a stochastic path in order for SpineOpt to know what scenario to use when: [stochastic structure](https://spine-tools.github.io/SpineOpt.jl/latest/advanced_concepts/stochastic_framework/). We are working on a tutorial that shows this hands on but it will be in the works for a while.

**Comment by SeanByrne08** [comment]
I added a map to unit_capacity in the 'PV' unit which contained different scenarios. I did it without attaching the stochatsic structure to the node. When i ran the code the output from my PV unit did not read the map and all the child scenarios contained the parent time series. 
![image](https://github.com/spine-tools/SpineOpt.jl/assets/141043201/facbc2fe-bf5c-440e-abd4-0958b5841107)
![image](https://github.com/spine-tools/SpineOpt.jl/assets/141043201/f41810b4-1f2a-4a1e-b2bf-d4cd88cc7640)

**Comment by SeanByrne08** [comment]
I have figured out that part of the probelm was the fact that i had model_default_stochastic_structure relationship class between EV flexibility and the deterministic stochastic structure. I have now attached relationships between the stchastic structures and the units/nodes individually
